services/data/src/react/hooks/useDataEngine.ts

```diff
@@ -1,8 +1,14 @@
-import { useContext } from 'react'
+import { useContext, useState } from 'react'
+import { DataEngine } from '../../engine'
 import { DataContext } from '../context/DataContext'
 
-export const useDataEngine = () => {
-    const context = useContext(DataContext)
+function useConst<T>(initializer: () => T): T {
+    const [value] = useState(initializer)
+    return value
+}
```

Comment:
```suggestion
function useInstanceVariable<T>(initializer: () => T, dependencies: any[]): T {
    const [value, setValue] = useState(initializer)
    const initializedRef = useRef(false)
    useEffect(() => {
        if (!initializedRef.current) {
            initializedRef.current = true
            return
        }
        setValue(initializer())
    }, dependencies)
    return value
}
```

Reply:
Maybe something like this instead, to support updates when deps change?

Reply:
Note however that this will cause double-renders on dependency updates...

Reply:
`useInstanceVariable` would be used in `DataProvider` and `CustomDataProvider` instead of in `useDataEngine` right?

Reply:
Sounds a lot like we could just use `useMemo` instead of `useConst`?

Reply:
@Mohammer5 this is super old... but no, there's a difference because `useConst` guarantees referential integrity while `useMemo` might re-initialize in certain situations since it is only considered a performance optimization.

Reply:
Yeah I also figured out the other day that `useMemo` is the wrong tool.
For simplicity's sake, I think the following would be the most straight forward:

```js
export const useDataEngine = (): DataEngine => {
    const context = useContext(DataContext)
    const engine = useState<DataEngine>(() => context.engine)
    return engine
}
```

This would remove the extra abstraction entirely as `useConst` is basically syntactic sugar for `useState(() => .....)`.

Reply:
```js
function useDataEngine = (): DataEngine => {
    const context = useContext(DataContext)
    const [instance, setInstance] = useState(context.engine)
    
   useEffect(() => {
        setInstance()
    }, )

    return value
}
```